File: src/muzik/scan.py
"""Scan music library and extract metadata from audio tags."""

import hashlib
from collections import defaultdict
from pathlib import Path

from mutagen import File as MutagenFile
from tqdm import tqdm

from muzik.db import (
    get_db,
    get_or_create_albums,
    get_or_create_tracks,
    upsert_albums,
    upsert_tracks,
)

AUDIO_EXTENSIONS = {".mp3", ".flac", ".ogg", ".opus", ".m4a", ".wma", ".wav", ".aiff"}
BATCH_SIZE = 500


def _track_id(filepath: str) -> str:
    return hashlib.sha256(filepath.encode()).hexdigest()[:16]


def _album_id(artist: str, album: str) -> str:
    key = f"{artist.lower().strip()}||{album.lower().strip()}"
    return hashlib.sha256(key.encode()).hexdigest()[:16]


def _get_tag(audio: MutagenFile, keys: list[str], default: str = "") -> str:
    """Extract a tag value, trying multiple key names for format compatibility."""
    for key in keys:
        val = audio.get(key)
        if val:
            # mutagen returns lists for most tag types
            if isinstance(val, list):
                return str(val[0])
            return str(val)
    return default


def _get_int_tag(audio: MutagenFile, keys: list[str], default: int = 0) -> int:
    raw = _get_tag(audio, keys, "")
    if not raw:
        return default
    # Handle "3/12" style track numbers
    raw = raw.split("/")[0].strip()
    try:
        return int(raw)
    except ValueError:
        return default


def _extract_metadata(filepath: Path) -> dict | None:
    """Extract metadata from a single audio file."""
    try:
        audio = MutagenFile(str(filepath), easy=True)
    except Exception:
        return None

    if audio is None:
        return None

    artist = _get_tag(audio, ["artist", "albumartist", "performer"])
    album_artist = _get_tag(audio, ["albumartist", "artist", "performer"])
    album = _get_tag(audio, ["album"])
    title = _get_tag(audio, ["title"])
    genre = _get_tag(audio, ["genre"])
    year = _get_int_tag(audio, ["date", "year", "originaldate"])
    track_num = _get_int_tag(audio, ["tracknumber"])

    # Fallback: infer artist/album from directory structure (Artist/Album/track.ext)
    if not artist and filepath.parent.parent.name:
        artist = filepath.parent.parent.name
    if not album and filepath.parent.name:
        album = filepath.parent.name
    if not title:
        title = filepath.stem

    return {
        "id": _track_id(str(filepath)),
        "title": title,
        "artist": artist,
        "album_artist": album_artist or artist,
        "album": album,
        "genre": genre,
        "year": year,
        "track_num": track_num,
        "filepath": str(filepath),
    }


def scan_library(library_path: Path) -> tuple[int, int]:
    """Scan a music library directory and store metadata in LanceDB.

    Returns (track_count, album_count).
    """
    # Collect all audio files
    audio_files = [
        p for p in library_path.rglob("*")
        if p.suffix.lower() in AUDIO_EXTENSIONS and p.is_file()
    ]

    if not audio_files:
        raise FileNotFoundError(f"No audio files found in {library_path}")

    db = get_db()
    tracks_table = get_or_create_tracks(db)

    # Extract metadata and batch-upsert
    batch: list[dict] = []
    total_tracks = 0
    album_data: dict[str, dict] = {}

    for filepath in tqdm(audio_files, desc="Scanning tracks", unit="file"):
        meta = _extract_metadata(filepath)
        if meta is None:
            continue

        batch.append(meta)
        total_tracks += 1

        # Collect album info
        aid = _album_id(meta["artist"], meta["album"])
        if aid not in album_data:
            album_data[aid] = {
                "id": aid,
                "artist": meta["album_artist"] or meta["artist"],
                "album": meta["album"],
                "year": meta["year"],
                "genres": defaultdict(int),
                "track_count": 0,
            }
        album_data[aid]["track_count"] += 1
        if meta["genre"]:
            album_data[aid]["genres"][meta["genre"]] += 1

        if len(batch) >= BATCH_SIZE:
            upsert_tracks(tracks_table, batch)
            batch.clear()

    if batch:
        upsert_tracks(tracks_table, batch)

    # Build and upsert albums
    albums_table = get_or_create_albums(db)
    album_rows = []
    for info in album_data.values():
        genres = info["genres"]
        # Sort genres by frequency, join
        sorted_genres = sorted(genres, key=genres.get, reverse=True)  # type: ignore[arg-type]
        album_rows.append({
            "id": info["id"],
            "artist": info["artist"],
            "album": info["album"],
            "year": info["year"],
            "genres": ", ".join(sorted_genres),
            "track_count": info["track_count"],
        })

    # Batch upsert albums
    for i in range(0, len(album_rows), BATCH_SIZE):
        upsert_albums(albums_table, album_rows[i : i + BATCH_SIZE])

    return total_tracks, len(album_data)

File: src/scan.rs
use std::collections::HashMap;
use std::path::Path;

use anyhow::{Context, Result, bail};
use indicatif::{ProgressBar, ProgressStyle};
use lofty::prelude::*;
use lofty::probe::Probe;
use sha2::{Digest, Sha256};
use walkdir::WalkDir;

use crate::db;
use crate::models::{AUDIO_EXTENSIONS, Album, Track};

fn track_id(filepath: &str) -> String {
    let hash = Sha256::digest(filepath.as_bytes());
    hex::encode(&hash[..8])
}

fn album_id(artist: &str, album: &str) -> String {
    let key = format!("{}||{}", artist.to_lowercase().trim(), album.to_lowercase().trim());
    let hash = Sha256::digest(key.as_bytes());
    hex::encode(&hash[..8])
}

fn get_tag_string(tag: &dyn Accessor, getter: fn(&dyn Accessor) -> Option<std::borrow::Cow<'_, str>>) -> String {
    getter(tag).map_or_else(String::new, |v| v.to_string())
}

fn extract_metadata(filepath: &Path) -> Option<Track> {
    let tagged_file = Probe::open(filepath)
        .ok()?
        .read()
        .ok()?;

    let tag = tagged_file.primary_tag().or_else(|| tagged_file.first_tag())?;

    let artist = get_tag_string(tag, |t| t.artist());
    let album = get_tag_string(tag, |t| t.album());
    let title = get_tag_string(tag, |t| t.title());
    let genre = get_tag_string(tag, |t| t.genre());
    let year = tag.year().map_or(0, u32::cast_signed);
    let track_num = tag.track().map_or(0, u32::cast_signed);

    // Get album_artist from ItemKey
    let album_artist = tag
        .get_string(&lofty::tag::ItemKey::AlbumArtist)
        .unwrap_or(&artist)
        .to_string();

    let filepath_str = filepath.to_string_lossy().to_string();

    // Fallback: infer from directory structure
    let final_artist = if artist.is_empty() {
        filepath
            .parent()
            .and_then(|p| p.parent())
            .map_or_else(String::new, |p| p.file_name().map_or_else(String::new, |n| n.to_string_lossy().to_string()))
    } else {
        artist
    };
    let final_album = if album.is_empty() {
        filepath
            .parent()
            .map_or_else(String::new, |p| p.file_name().map_or_else(String::new, |n| n.to_string_lossy().to_string()))
    } else {
        album
    };
    let final_title = if title.is_empty() {
        filepath
            .file_stem()
            .map_or_else(String::new, |s| s.to_string_lossy().to_string())
    } else {
        title
    };

    Some(Track {
        id: track_id(&filepath_str),
        title: final_title,
        artist: final_artist,
        album_artist: if album_artist.is_empty() {
            "Unknown".to_string()
        } else {
            album_artist
        },
        album: final_album,
        genre,
        year,
        track_num,
        filepath: filepath_str,
        metadata_vec: vec![0.0; crate::models::METADATA_DIM],
        audio_vec: vec![0.0; crate::models::AUDIO_DIM],
    })
}

/// Scan a music library directory and store metadata in `LanceDB`.
///
/// Returns `(track_count, album_count)`.
pub fn scan_library(library_path: &Path, quiet: bool) -> Result<(usize, usize)> {
    if !library_path.is_dir() {
        bail!("{} is not a directory", library_path.display());
    }

    // Collect audio files
    let audio_files: Vec<_> = WalkDir::new(library_path)
        .into_iter()
        .filter_map(std::result::Result::ok)
        .filter(|e| e.file_type().is_file())
        .filter(|e| {
            e.path()
                .extension()
                .and_then(|ext| ext.to_str())
                .is_some_and(|ext| AUDIO_EXTENSIONS.contains(&ext.to_lowercase().as_str()))
        })
        .map(walkdir::DirEntry::into_path)
        .collect();

    if audio_files.is_empty() {
        bail!("No audio files found in {}", library_path.display());
    }

    let pb = if quiet {
        ProgressBar::hidden()
    } else {
        let pb = ProgressBar::new(audio_files.len() as u64);
        pb.set_style(
            ProgressStyle::default_bar()
                .template("[{elapsed_precise}] {bar:40} {pos}/{len} {msg}")
                .context("progress bar template")?,
        );
        pb.set_message("Scanning tracks");
        pb
    };

    let mut db = db::open_db()?;
    let mut all_tracks: Vec<Track> = Vec::with_capacity(audio_files.len());
    let mut album_data: HashMap<String, AlbumAccumulator> = HashMap::new();

    for filepath in &audio_files {
        pb.inc(1);
        let Some(track) = extract_metadata(filepath) else {
            continue;
        };

        // Accumulate album info
        let aid = album_id(&track.artist, &track.album);
        let entry = album_data.entry(aid.clone()).or_insert_with(|| AlbumAccumulator {
            id: aid,
            artist: track.album_artist.clone(),
            album: track.album.clone(),
            year: track.year,
            genres: HashMap::new(),
            track_count: 0,
        });
        entry.track_count += 1;
        if !track.genre.is_empty() {
            *entry.genres.entry(track.genre.clone()).or_insert(0) += 1;
        }

        all_tracks.push(track);
    }

    let total_tracks = all_tracks.len();
    db::write_all_tracks(&mut db, &all_tracks)?;

    // Build and write albums
    let albums: Vec<Album> = album_data
        .into_values()
        .map(|acc| {
            let mut sorted_genres: Vec<_> = acc.genres.into_iter().collect();
            sorted_genres.sort_by(|a, b| b.1.cmp(&a.1));
            let genres_str = sorted_genres
                .into_iter()
                .map(|(g, _)| g)
                .collect::<Vec<_>>()
                .join(", ");
            Album {
                id: acc.id,
                artist: acc.artist,
                album: acc.album,
                year: acc.year,
                genres: genres_str,
                track_count: acc.track_count,
                metadata_vec: vec![0.0; crate::models::METADATA_DIM],
                audio_vec: vec![0.0; crate::models::AUDIO_DIM],
            }
        })
        .collect();

    let album_count = albums.len();
    db::write_all_albums(&mut db, &albums)?;

    pb.finish_and_clear();
    Ok((total_tracks, album_count))
}

struct AlbumAccumulator {
    id: String,
    artist: String,
    album: String,
    year: i32,
    genres: HashMap<String, i32>,
    track_count: i32,
}
